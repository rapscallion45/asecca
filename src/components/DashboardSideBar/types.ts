/*
 * Dashboard Sidebar type definitions
 *
 * These type definitions are used in the applications Dashboard Sidebar component.
 */

/**
 * Nav Bar sidebar on close callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback IDashboardSidebarOnCloseCallback
 * @returns {void} - no return value
 */
export interface IDashboardSidebarOnCloseCallback {
  (): void;
}
