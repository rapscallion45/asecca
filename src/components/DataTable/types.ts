/*
 * Data Table type definitions
 *
 * These type definitions are used in the applications data table component.
 * Only make changes to this file if the corresponding change is made in the
 * Data Table component(s)
 */

/**
 * Data Table Row Cell Value types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 * @memberof DataTable
 *
 * @typedef DataTableRowCellValue
 */
export type DataTableRowCellValue =
  | string
  | Array<string>
  | boolean
  | null
  | undefined;

/**
 * Data Table Column types
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 * @memberof DataTable
 *
 * @typedef DataTableColumnType
 */
export type DataTableColumnType = 'string' | 'currency' | 'select' | 'check';

/**
 * Data Table Column properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableColumn
 * @prop {string} label - data table column name
 * @prop {string} key - data table column data identifier
 * @prop {DataTableColumnType} type - data table column type
 */
export interface IDataTableColumn {
  label: string;
  key: string;
  type: DataTableColumnType;
}

/**
 * Data Table Row properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @typedef IDataTableRow
 * @prop {string} label - data table row name
 */
export interface IDataTableRow {
  label: string;
}

/**
 * Data Table Cell Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCellValueCallback
 * @param {DataTableRowCellValue} value - updated cell value, can be null
 * @param {string} colKey - cell column data identifier
 * @param {number} rowIdx - cell row index in data table
 * @returns {void} - no return value
 */
export interface IDataTableEditCellValueCallback {
  (value: DataTableRowCellValue, colKey: string, rowIdx: number): void;
}

/**
 * Data Table Cell Value Retrieval callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableGetCellValueCallback
 * @param {number} rowIdx - cell row index in data table
 * @param {IDataTableColumn} column - cell column in data table
 * @returns {DataTableRowCellValue} - cell value, can be null or undefined
 */
export interface IDataTableGetCellValueCallback {
  (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue;
}

/**
 * Data Table Currency Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.0
 *
 * @callback IDataTableEditCurrencyCellValueCallback
 * @param {string | null} value - updated cell value
 * @returns {void} - no return value
 */
export interface IDataTableEditCurrencyCellValueCallback {
  (value: string | null): void;
}

/**
 * Data Table Checkbox Cell Value Edit callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.13
 *
 * @callback IDataTableEditCheckboxCellValueCallback
 * @param {boolean | null} value - updated cell boolean value
 * @returns {void} - no return value
 */
export interface IDataTableEditCheckboxCellValueCallback {
  (value: boolean | null): void;
}
