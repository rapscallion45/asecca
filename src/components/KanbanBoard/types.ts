/*
 * Kanban Board type definitions
 *
 * These type definitions are used in the applications kanban board component.
 * Only make changes to this file if the corresponding change is made in the
 * Kanban Board component(s)
 */

/**
 * Kanban Board Column On Drop callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.7
 *
 * @callback IKanbanBoardColumnOnDropCallback
 * @param {number} colIndex - column index of task
 * @param {number} prevColIndex - previous column index of task
 * @param {number} taskIndex - index of task
 * @returns {void} - no return value
 */
export interface IKanbanBoardColumnOnDropCallback {
  (colIndex: number, prevColIndex: number, taskIndex: number): void;
}
