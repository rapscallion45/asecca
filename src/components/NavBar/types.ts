/*
 * Nav Bar type definitions
 *
 * These type definitions are used in the applications Nav Bar component.
 */

/**
 * Nav Bar sidebar open callback function typedef
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 *
 * @callback INavBarOnSidebarOpenCallback
 * @returns {void} - no return value
 */
export interface INavBarOnSidebarOpenCallback {
  (): void;
}
