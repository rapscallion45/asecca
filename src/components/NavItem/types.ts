/*
 * Nav Item type definitions
 *
 * These type definitions are used in the applications Nav Item component.
 */

/**
 * Nav Item configuration properties
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.1
 * @memberof NavItem
 *
 * @typedef INavItemConfig
 * @prop {string} title - nav item title text
 * @prop {string} path - nav item URL path
 * @prop {any} icon - nav item icon
 * @prop {boolean} newTab - nav item link opens in new tab
 * @prop {Array<INavItemConfig>} children - nav item child components
 */
export interface INavItemConfig {
  title: string;
  path?: string;
  icon?: any;
  newTab?: boolean;
  children?: Array<INavItemConfig>;
}
