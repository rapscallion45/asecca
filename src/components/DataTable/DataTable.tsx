import { FC, Fragment } from 'react';
import { styled } from '@mui/material/styles';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import Box from '@mui/material/Box';
import CircularProgress from '@mui/material/CircularProgress';
import Typography from '@mui/material/Typography';
import ErrorIcon from '@mui/icons-material/Error';
import DataTableRow from './DataTableRow/DataTableRow';
import ClientOnly from '../ClientOnly/ClientOnly';

/* table head cell stylings */
const StyledTableHeadCell = styled(TableCell)(({ theme }) => ({
  /* table head colors */
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: theme.palette.secondary.main,
    color: theme.palette.primary.contrastText,
    fontSize: 16,
    fontWeight: 400,
  },
}));

interface DataTableProps {
  name: string;
  editColName: string;
  columns: any;
  rows: any;
  isLoading?: boolean;
  isError?: boolean;
}

/* Data Table */
/* ========== */
const DataTable: FC<DataTableProps> = (props) => {
  const {
    name,
    editColName,
    columns,
    rows,
    isLoading = false,
    isError = false,
  } = props;

  /* ClientOnly used to not allow tables to SSR */
  return (
    <ClientOnly>
      <TableContainer component={Paper}>
        <Table aria-label={`${name} table`}>
          <TableHead>
            <TableRow>
              {/* map passed columns to table headers row */}
              {columns.map((column: any) => (
                <StyledTableHeadCell key={column.key} align="center">
                  {column.label}
                </StyledTableHeadCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            <>
              {!isLoading && !isError && (
                <>
                  {/* map passed rows */}
                  {rows.map((row: any) => (
                    <Fragment key={row.name}>
                      <DataTableRow
                        row={row}
                        columns={columns}
                        /* we need the edit col to get 'Prevailing' value */
                        editCol={columns.find(
                          (col: any) => editColName === col.label
                        )}
                      />
                    </Fragment>
                  ))}
                </>
              )}
              {!isLoading && isError && (
                <Box
                  display="flex"
                  flexDirection="column"
                  py={7}
                  alignItems="center"
                >
                  <ErrorIcon color="error" fontSize="large" />
                  <Typography variant="h6" mt={2}>
                    Error loading the requested data
                  </Typography>
                </Box>
              )}
              {isLoading && (
                <Box
                  display="flex"
                  flexDirection="column"
                  py={7}
                  alignItems="center"
                >
                  <CircularProgress />
                  <Typography variant="h6" mt={2}>
                    Loading...
                  </Typography>
                </Box>
              )}
            </>
          </TableBody>
        </Table>
      </TableContainer>
    </ClientOnly>
  );
};

export default DataTable;
