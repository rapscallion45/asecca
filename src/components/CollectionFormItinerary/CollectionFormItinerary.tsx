import { FC, useState, useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState, AppDispatch } from '@/redux/store';
import {
  Card,
  CardHeader,
  CardContent,
  Box,
  Button,
  IconButton,
} from '@mui/material';
import { LoadingButton } from '@mui/lab';
import DeleteIcon from '@mui/icons-material/Delete';
import CategoryIcon from '@mui/icons-material/Category';
import {
  ICollectionFormItineraryData,
  INewAssetCategoryDataPayload,
} from '@/lib/api/api-types';
import {
  saveByCollectionId as saveItineraryByCollectionId,
  resetItinerary,
  editItinerary,
  addItinerary,
  deleteItinerary,
  // fetchTypes,
  saveNewAssetCategory,
} from '@/redux/slices/collectionFormItinerarySlice';
import DataTable from '@/components/DataTable/DataTable';
import {
  DataTableRowCellValue,
  IDataTableColumn,
} from '@/components/DataTable/types';
import FormModal from '@/modals/FormModal/FormModal';
import NewAssetCategoryModalModal from '../NewAssetCategoryModal/NewAssetCategoryModal';
import columns from './collectionFormItineraryTableColumns';

/**
 * Collection Form Itinerary Props
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 *
 * @typedef ICollectionFormItineraryProps
 * @prop {string} collectionId - ID string of Collection for table data API call
 */
interface ICollectionFormItineraryProps {
  collectionId: string;
}

/**
 * Collection Form Itinerary
 *
 * Presents the Collection Form Itinerary table to the user, populated with data
 * fetched from API: /api/collection/itinerary/api/itinerary
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 *
 * @component
 * @param {ICollectionFormItineraryProps} props - component props
 * @returns {FC} - collection form itinerary table functional component
 */
const CollectionFormItinerary: FC<ICollectionFormItineraryProps> = (props) => {
  const { collectionId } = props;

  /* shorthand helper for dispatching redux actions */
  const dispatch = useDispatch<AppDispatch>();

  /* get collection form itinerary data held in redux state */
  const {
    data,
    loading,
    error,
    saving,
    edited,
    savingNewAssetCategory,
    assetCategories,
  } = useSelector((state: AppState) => state.collectionFormItinerary);

  /* on first load, fetch the asset categories */
  // useEffect(() => {
  //   dispatch(fetchTypes());
  // }, [dispatch]);

  /**
   * Table column list
   *
   * Local state of columns list for available itinerary and asset categories
   *
   * @since 0.0.15
   *
   * @constant
   */
  const [colList, setColList] = useState<Array<IDataTableColumn>>(columns);

  /* Whenever Logisitcs data updates, update column list */
  useEffect(() => {
    setColList(
      columns.map((col: IDataTableColumn) => {
        if (col.key === 'asset_category')
          return {
            ...col,
            selectOptions: assetCategories,
          };
        return col;
      })
    );
  }, [data, assetCategories]);

  /**
   * Handles the saving of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.15
   *
   * @method
   */
  const handleSave = useCallback(() => {
    dispatch(
      saveItineraryByCollectionId({
        data: {
          collection: collectionId,
          items: data,
        },
      })
    );
  }, [collectionId, data, dispatch]);

  /**
   * Handles the resetting of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   */
  const handleReset = useCallback(() => {
    dispatch(resetItinerary());
  }, [dispatch]);

  /**
   * Handles the update of the table data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.13
   *
   * @method
   * @param {DataTableRowCellValue} value - value to be used in update
   * @param {string} colKey - column key to be updated
   * @param {number} rowIdx - table row index to be updated
   */
  const handleEditCellValue = useCallback(
    (value: DataTableRowCellValue, colKey: string, rowIdx: number) => {
      dispatch(editItinerary({ value, colKey, rowIdx }));
    },
    [dispatch]
  );

  /**
   * Handles any required logic when determining a cell's display value
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.15
   *
   * @method
   * @param {number} rowIdx - table row index to get value from
   * @param {IDataTableColumn} column - column to get value from
   * @return {DataTableRowCellValue} - cell value, can be null or undefined
   */
  const handleGetCellValue = useCallback(
    (rowIdx: number, column: IDataTableColumn): DataTableRowCellValue =>
      data[rowIdx][column.key as keyof ICollectionFormItineraryData],
    [data]
  );

  /**
   * Handles addition of table row for new Itinerary type
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.15
   *
   * @method
   */
  const handleAddRow = useCallback(() => dispatch(addItinerary()), [dispatch]);

  /**
   * Handles deletion of table row for passed row index
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.15
   *
   * @method
   * @param {number} rowIdx - table row index to be deleted
   */
  const handleDeleteRow = useCallback(
    (rowIdx: number) => dispatch(deleteItinerary({ rowIdx })),
    [dispatch]
  );

  /**
   * Handles requests for the action component for specific column
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.15
   *
   * @method
   * @param {string} colKey - table column key of the action cell
   * @param {number} rowIdx - table row index of the action cell
   * @returns {ReactNode} - action component to present
   */
  const getActionComponent = (colKey: string, rowIdx: number) => (
    <IconButton aria-label="delete" onClick={() => handleDeleteRow(rowIdx)}>
      <DeleteIcon />
    </IconButton>
  );

  /**
   * Handles save callback of the New Asset Category modal
   *
   * If new asset category has been added, this should trigger new fetch
   * of the asset category data
   *
   * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
   * @since 0.0.20
   *
   * @method
   */
  const handleSaveNewAssetCategory = (
    newAssetData: INewAssetCategoryDataPayload
  ) => {
    dispatch(saveNewAssetCategory({ data: newAssetData }));
  };

  return (
    <Card>
      <CardHeader title="Itinerary" />
      <CardContent sx={{ pt: 0 }}>
        <DataTable
          name="collection form itinerary"
          /* filter table columns by current itinerary type */
          columns={colList}
          /* can edit all cells in itinerary table */
          editableColLabels={colList.map((col: IDataTableColumn) => col.label)}
          /* build table row props from itinerary data */
          rows={data?.map((itinerary: ICollectionFormItineraryData) => ({
            label: itinerary.asset_category,
          }))}
          isLoading={loading}
          error={error}
          editCellValueCallback={handleEditCellValue}
          getCellValueCallback={handleGetCellValue}
          getActionComponent={getActionComponent}
        />
        <Box
          sx={{
            marginTop: 2,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'left',
          }}
        >
          <Button
            color="secondary"
            variant="contained"
            onClick={handleAddRow}
            disabled={saving || loading}
            sx={{ minWidth: '80.61px', mr: 2 }}
          >
            Add +
          </Button>
          <FormModal
            triggerBtn={{
              type: 'menu',
              // @ts-ignore
              icon: CategoryIcon,
              iconStyle: { marginRight: '10px' },
              text: 'Add Asset Category +',
              closeMenu: () => {},
            }}
            title="New Asset Category"
          >
            <NewAssetCategoryModalModal
              saving={savingNewAssetCategory}
              handleSaveCallback={handleSaveNewAssetCategory}
            />
          </FormModal>
        </Box>
        <Box
          sx={{
            marginTop: 2,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'left',
          }}
        >
          <Button
            color="secondary"
            variant="outlined"
            onClick={handleReset}
            disabled={saving || loading || !edited}
            sx={{ backgroundColor: 'common.white' }}
          >
            Reset
          </Button>
          <LoadingButton
            color="secondary"
            variant="contained"
            onClick={handleSave}
            disabled={saving || loading || !edited}
            loading={saving}
            sx={{ ml: 2 }}
          >
            Save
          </LoadingButton>
        </Box>
      </CardContent>
    </Card>
  );
};

export default CollectionFormItinerary;
