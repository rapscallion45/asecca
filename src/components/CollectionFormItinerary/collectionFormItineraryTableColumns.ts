import { IDataTableColumn } from '@/components/DataTable/types';

/**
 * Collection Form Itinerary table column defintions
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.15
 * @memberof CollectionFormItinerary
 *
 * @constant
 * @type {Array<IDataTableColumn>}
 */
const collectionFormItineraryTableColumns: Array<IDataTableColumn> = [
  {
    label: '',
    key: 'delete_row',
    type: 'action',
  },
  {
    label: 'Asset Category',
    key: 'asset_category',
    type: 'select',
    selectOptions: [] /* select options retreived from API at runtime */,
  },
  {
    label: 'Quantity',
    key: 'quantity',
    type: 'numerical',
    nullDisallowed: true,
  },
  {
    label: 'Packing Details',
    key: 'packing_details',
    type: 'text',
  },
];

export default collectionFormItineraryTableColumns;
