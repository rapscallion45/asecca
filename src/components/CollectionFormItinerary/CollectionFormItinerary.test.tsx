import React from 'react';
import { Provider } from 'react-redux';
import {
  fireEvent,
  render,
  screen,
  waitFor,
  act,
} from '@testing-library/react';
import renderer from 'react-test-renderer';
import { fetchByCollectionId as fetchCollectionFormItineraryByCollectionId } from '@/redux/slices/collectionFormItinerarySlice';
import { ICollectionFormItineraryData } from '@/lib/api/api-types';
import store from '../../redux/store';
import columns from './collectionFormItineraryTableColumns';
import CollectionFormItinerary from './CollectionFormItinerary';
import collectionFormItineraryDataMock from '../../../__mocks__/CollectionForm/collectionFormItineraryDataMock';
import collectionFormItineraryAssetCategoryDataMock from '../../../__mocks__/CollectionForm/collectionFormItineraryAssetCategoryDataMock';
import { IDataTableColumn } from '../DataTable/types';

/* default test query ID */
const query: string = '66135000001760012';

/* chosen data row index to check during tests */
const testDataRowIdx = 1;

/* test Logisitics Type index to check during tests */
const testTypeIdx = 0;

/**
 * Collection Form Itinerary Unit Tests
 *
 * @author Carl Scrivener {@link https://github.com/rapscallion45 GitHub}
 * @since 0.0.21
 */
describe('Collection Form Itinerary', () => {
  it('Renders correctly', async () => {
    /** Arrange */
    store.dispatch(
      fetchCollectionFormItineraryByCollectionId({
        collectionId: query,
      })
    );

    /** perform snapshot test */
    const tree = renderer
      .create(
        <Provider store={store}>
          <CollectionFormItinerary collectionId={query} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Table Callbacks', () => {
    beforeAll(() => jest.spyOn(window, 'fetch'));

    it('Should call API with updated values when value updated and save button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormItineraryByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Assert - get API to have been called */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(1);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        `/api/collection/itinerary/api/itinerary?collection=${query}`,
        expect.objectContaining({
          method: 'GET',
          headers: { 'Content-Type': 'application/json' },
        })
      );
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormItinerary collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'delete_row')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormItineraryDataMock.forEach(
        (mockData: ICollectionFormItineraryData) => {
          expect(
            screen.getByDisplayValue(mockData.asset_category)
          ).toBeInTheDocument();
        }
      );
      /** Act - click on dropdown, update value, click Save button */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormItineraryDataMock[testDataRowIdx].asset_category
          ),
          {
            target: {
              value: collectionFormItineraryAssetCategoryDataMock[testTypeIdx],
            },
          }
        );
      });
      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormItineraryAssetCategoryDataMock[testTypeIdx]
          )
        ).toBeInTheDocument();
      });
      /** Act - click Save button */
      act(() => {
        fireEvent(
          screen.getByText('Save'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - save API to have been called with data from state */
      await waitFor(() => {
        expect(window.fetch).toHaveBeenCalledTimes(2);
      });
      expect(window.fetch).toHaveBeenCalledWith(
        '/api/collection/itinerary/api/itinerary',
        expect.objectContaining({
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            collection: query,
            items: store.getState().collectionFormItinerary.data,
          }),
        })
      );
    });

    it('Should reset table data when cancel button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormItineraryByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormItinerary collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'delete_row')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormItineraryDataMock.forEach(
        (mockData: ICollectionFormItineraryData) => {
          expect(
            screen.getByDisplayValue(mockData.asset_category)
          ).toBeInTheDocument();
        }
      );
      /** Act - click on dropdown, update value */
      act(() => {
        fireEvent.change(
          screen.getByDisplayValue(
            collectionFormItineraryDataMock[testDataRowIdx].asset_category
          ),
          {
            target: {
              value: collectionFormItineraryAssetCategoryDataMock[testTypeIdx],
            },
          }
        );
      });
      /** Assert - wait for input to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormItineraryAssetCategoryDataMock[testTypeIdx]
          )
        ).toBeInTheDocument();
      });
      /** Act - click Reset button */
      act(() => {
        fireEvent(
          screen.getByText('Reset'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - table data should no longer hold updated values and be reset */
      await waitFor(() => {
        expect(
          screen.queryByDisplayValue(
            collectionFormItineraryAssetCategoryDataMock[testTypeIdx]
          )
        ).toBeNull();
      });
      expect(
        screen.getByDisplayValue(
          collectionFormItineraryDataMock[testDataRowIdx].asset_category
        )
      ).toBeInTheDocument();
    });

    it('Should add row to table when "Add +" button clicked', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormItineraryByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormItinerary collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'delete_row')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormItineraryDataMock.forEach(
        (mockData: ICollectionFormItineraryData) => {
          expect(
            screen.getByDisplayValue(mockData.asset_category)
          ).toBeInTheDocument();
        }
      );
      /** Act - click on "Add +" button */
      act(() => {
        fireEvent(
          screen.getByText('Add +'),
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - wait for data to update */
      await waitFor(() => {
        expect(
          screen.getByDisplayValue(
            collectionFormItineraryAssetCategoryDataMock[testTypeIdx]
          )
        ).toBeInTheDocument();
      });
    });

    it('Should delete row from table when delete icon button clicked', async () => {
      /** Arrange */
      const deleteRowIdx = 0;
      act(() => {
        store.dispatch(
          fetchCollectionFormItineraryByCollectionId({
            collectionId: query,
          })
        );
      });
      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormItinerary collectionId={query} />
        </Provider>
      );
      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'delete_row')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
      /* check data has been rendered */
      collectionFormItineraryDataMock.forEach(
        (mockData: ICollectionFormItineraryData) => {
          expect(
            screen.getByDisplayValue(mockData.asset_category)
          ).toBeInTheDocument();
        }
      );
      /** Act - click on delete icon button of first row */
      act(() => {
        fireEvent(
          screen.getAllByTestId('DeleteIcon')[deleteRowIdx],
          new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
          })
        );
      });
      /** Assert - wait for data to update */
      await waitFor(() => {
        expect(
          screen.queryByDisplayValue(
            collectionFormItineraryDataMock[testDataRowIdx].asset_category
          )
        ).toBeNull();
      });
    });
  });

  describe('Table Columns', () => {
    it('Should render correctly for passed permission', async () => {
      /** Arrange */
      act(() => {
        store.dispatch(
          fetchCollectionFormItineraryByCollectionId({
            collectionId: query,
          })
        );
      });

      /** Act - render the test components */
      render(
        <Provider store={store}>
          <CollectionFormItinerary collectionId={query} />
        </Provider>
      );

      /** Assert - check that table columns have been rendered */
      await waitFor(() => {
        /** After all state updates have completed */
        columns.forEach((column: IDataTableColumn) => {
          if (column.key !== 'delete_row')
            expect(screen.getByText(column.label)).toBeInTheDocument();
        });
      });
    });
  });
});
