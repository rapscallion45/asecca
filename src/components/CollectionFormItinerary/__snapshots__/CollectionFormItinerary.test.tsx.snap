// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Form Itinerary Renders correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
>
  <div
    className="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
  >
    <div
      className="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
    >
      <span
        className="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
      >
        Itinerary
      </span>
    </div>
  </div>
  <div
    className="MuiCardContent-root css-1hizqsf-MuiCardContent-root"
  >
    <div
      className="MuiBox-root css-kh5ii8"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-19c6wdr-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Add +
      </button>
      <div
        style={
          {
            "display": "inline",
          }
        }
      >
        <li
          className="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="menuitem"
          tabIndex={-1}
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CategoryIcon"
            focusable="false"
            style={
              {
                "marginRight": "10px",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="m12 2-5.5 9h11z"
            />
            <circle
              cx="17.5"
              cy="17.5"
              r="4.5"
            />
            <path
              d="M3 13.5h8v8H3z"
            />
          </svg>
          Add Asset Category +
        </li>
      </div>
    </div>
    <div
      className="MuiBox-root css-kh5ii8"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium Mui-disabled MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1ex1hsz-MuiButtonBase-root-MuiButton-root"
        disabled={true}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Reset
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fddvdp-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        disabled={true}
        id=":r1:"
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
